zson/formatter.go

```diff
@@ -206,6 +208,174 @@ func (f *Formatter) formatValue(indent int, typ zed.Type, bytes zcode.Bytes, par
+		f.startColor(color.Gray(160))
+		f.build(zed.PrimitiveName(typ))
+		f.endColor()
+	case zed.TypeValueNameDef:
+		name, tv := zed.DecodeName(tv)
+		if tv == nil {
+			f.truncTypeValueErr()
+			return nil
+		}
+		f.build(name)
+		f.build("=")
+		return f.formatTypeValue(indent, tv)
+	case zed.TypeValueNameRef:
+		name, tv := zed.DecodeName(tv)
+		if tv == nil {
+			f.truncTypeValueErr()
+			return nil
+		}
+		f.build(name)
+		return tv
+	case zed.TypeValueRecord:
+		f.build("{")
+		var n int
+		n, tv = zed.DecodeLength(tv)
+		if tv == nil {
+			f.truncTypeValueErr()
+			return nil
+		}
+		if n == 0 {
+			f.build("}")
+			return tv
+		}
+		sep := f.newline
+		indent += f.tab
+		for k := 0; k < n; k++ {
+			f.build(sep)
+			var name string
+			name, tv = zed.DecodeName(tv)
+			if tv == nil {
+				f.truncTypeValueErr()
+				return nil
+			}
+			f.indent(indent, QuotedName(name))
+			f.build(":")
+			if f.tab > 0 {
+				f.build(" ")
+			}
+			tv = f.formatTypeValue(indent, tv)
+			sep = "," + f.newline
+		}
+		f.build(f.newline)
+		f.indent(indent-f.tab, "}")
+	case zed.TypeValueArray:
+		tv = f.formatVectorTypeValue("[", indent, tv)
+	case zed.TypeValueSet:
+		tv = f.formatVectorTypeValue("|[", indent, tv)
+	case zed.TypeValueMap:
+		f.build("|{")
+		newline := f.newline
+		indent += f.tab
+		if n, itv := zed.DecodeLength(tv); n <= zed.IDNull {
+			n, _ = zed.DecodeLength(itv)
+			if n <= zed.IDNull {
+				// If key and value are both primitives don't indent.
+				indent -= f.tab
+				newline = ""
+			}
+		}
+		f.build(newline)
+		f.indent(indent, "")
+		tv = f.formatTypeValue(indent, tv)
+		f.build(":")
+		if f.tab > 0 {
+			f.build(" ")
+		}
+		tv = f.formatTypeValue(indent, tv)
+		f.build(newline)
+		if newline != "" {
+			f.indent(indent-f.tab, "}|")
+		} else {
+			f.build("}|")
+		}
+	case zed.TypeValueUnion:
+		f.build("(")
+		var n int
+		n, tv = zed.DecodeLength(tv)
+		if tv == nil {
+			f.truncTypeValueErr()
+			return nil
+		}
+		sep := f.newline
+		indent += f.tab
+		for k := 0; k < n; k++ {
+			f.build(sep)
+			f.indent(indent, "")
+			tv = f.formatTypeValue(indent, tv)
+			sep = "," + f.newline
+		}
+		f.build(f.newline)
+		f.indent(indent-f.tab, ")")
+	case zed.TypeValueEnum:
+		f.build("enum(")
+		var n int
+		n, tv = zed.DecodeLength(tv)
+		if tv == nil {
+			f.truncTypeValueErr()
+			return nil
+		}
+		for k := 0; k < n; k++ {
+			if k > 0 {
+				f.build(",")
+			}
+			var symbol string
+			symbol, tv = zed.DecodeName(tv)
+			if tv == nil {
+				return nil
+			}
+			f.build(QuotedName(symbol))
+		}
+		f.build(")")
+	case zed.TypeValueError:
+		f.startColor(color.Red)
+		f.build("error")
+		f.endColor()
+		f.build("(")
+		tv = f.formatTypeValue(indent, tv)
+		f.build(")")
+	}
+	return tv
+}
+
+func (f *Formatter) formatVectorTypeValue(token string, indent int, tv zcode.Bytes) zcode.Bytes {
+	f.build(token)
+	end := "]"
+	if token == "|[" {
+		end += "|"
+	}
+	if n, _ := zed.DecodeLength(tv); n <= zed.IDTypeComplex {
```

Comment:
Check for an error and call `Formatter.truncTypeValueErr`.

Reply:
No need, as this will get caught on the call to `formatTypeValue`.

Reply:
But this should be `n < zed.IDTypeComplex`